source/module_basis/module_nao/pswfc_radials.cpp

```diff
@@ -0,0 +1,308 @@
+    while((std::fabs(delta_norm) > conv_thr)&&(ir_ <= ir_max_))
+    {
+        ir_ = std::min(ir_ + delta_ir, ir_max_); // update ir_, but be careful not to exceed ir_max_
+        delta_norm = norm;
+        std::vector<double> rgrid_slice = std::vector<double>(rgrid.begin() + ir_min_, rgrid.begin() + ir_ + 1);
+        std::vector<double> rvalue_slice = std::vector<double>(rvalue.begin() + ir_min_, rvalue.begin() + ir_ + 1);
+        norm = radial_norm(rgrid_slice, rvalue_slice);
+        delta_norm = norm - delta_norm;
+        if(istep == 1) printf("%10d%12.2f%14.10f%18.10e\n", istep, rgrid[ir_], norm, delta_norm);
+        ++istep;
+    }
+    printf("%10d%12.2f%14.10f%18.10e\n", istep, rgrid[ir_], norm, delta_norm);
+
+    rvalue = std::vector<double>(rvalue.begin() + ir_min_, rvalue.begin() + ir_ + 1);
+    return rgrid[ir_max_];
+}
+
+void PswfcRadials::smooth(std::vector<double>& rgrid,
+                          std::vector<double>& rvalue,
+                          const double sigma)
+{
+    double prefactor = 1.0 / sqrt(2.0 * M_PI) / sigma;
+    double rmax = rgrid.back();
+    for(int ir = 0; ir != rgrid.size(); ++ir)
+    {
+        double delta_r = rgrid[ir] - rmax;
+        double smooth = prefactor * exp(-delta_r * delta_r / 2.0 / sigma / sigma);
+        rvalue[ir] *= (1 - smooth);
+    }
+}
+
+std::vector<double> PswfcRadials::pswfc_prepossess(std::map<std::pair<int, int>, std::vector<double>>& lzeta_rvalues)
+{
+    double nmax = 0.0;
+    for(auto it = lzeta_rvalues.begin(); it != lzeta_rvalues.end(); it++)
+    {
+        int l = it->first.first;
+        int iz = it->first.second;
+        std::vector<double> rvalue = it->second;
+        std::vector<double> rgrid = std::vector<double>(rvalue.size(), 0.0);
+        for(int ir = 0; ir < rvalue.size(); ir++)
+        {
+            rgrid[ir] = ir * 0.01;
+        }
+        double rcut_i = cut_to_convergence(rgrid, rvalue, 1e-6);
+        if(rvalue.size() > nmax) nmax = rvalue.size();
+        lzeta_rvalues[it->first] = rvalue; // stores in map
+    }
+    // generate rgrid
+    std::vector<double> rgrid = std::vector<double>(nmax, 0.0);
+    for(int ir = 0; ir < nmax; ir++)
+    {
+        rgrid[ir] = ir * 0.01;
+    }
+    // zero padding on rvalue
+    for(auto it = lzeta_rvalues.begin(); it != lzeta_rvalues.end(); it++)
+    {
+        int l = it->first.first;
+        int iz = it->first.second;
+        std::vector<double> rvalue = it->second;
+        std::vector<double> rvalue_padded = std::vector<double>(nmax, 0.0);
+        for(int ir = 0; ir < rvalue.size(); ir++)
+        {
+            rvalue_padded[ir] = rvalue[ir];
+        }
+        smooth(rgrid, rvalue_padded, 0.1); // smooth the radial function to avoid high frequency noise in FFT-spherical bessel transform
+        lzeta_rvalues[it->first] = rvalue_padded; // stores in map
+    }
+
+    return rgrid;
+}
+
+void PswfcRadials::read_upf_pswfc(std::ifstream& ifs, 
```

Comment:
read pseudopotential file. Presently because lcao can only support norm-conserving pseudopotential and most of this kind of pseudopotential is generated by ONCVPSP code.